### ISSUE TITLE
Adding string to the tikzpicture environment

### ISSUE THREAD

**Issue by t4qjXH8N** [matlab2tikz/matlab2tikz#393]
Situation: I manually added an extra axis with images and arrows to a regular dataplot, created with matlab2tikz. This manually added axis has its own grid, which makes it easy to position the images and the arrows. 

Problem: Everytime when I update my dataplot with matlab2tikz, the manually added axis is lost (of course). 

My approach: I created a text file containing the new axis environment only. After reading it into a string with MatLAB, I could not manage to add the string into the tikzpicture environment. Adding strings to the OPTIONS of the tikzpicture environment is, of course, possible but does not solve the problem.

Feature Request: Add an option to add strings to the tikzpicture environment.

Or are there any other ideas?

**Comment by egeerardyn** [comment]
After #670, I think the best option would be to pass some options in `UserData` of the HG object tree.

One of the points I previously marked as a downside, i.e. "it would riddle the code with checks just for handling the UserData.", can actually be used to our advantage, e.g. using an implementation like
```matlab
function [m2t, str] = handleObject(m2t, h, actualHandler)
    assert(isa(actualHandler,'function_handle'));

    UD = get(h, 'UserData');
    UD = sanitizeUserData(UD); % make sure the data is what we expect, remove all the rest
    extraOptions = getExtraOptions(UD);
    [m2t, str] = actualHandler(m2t, h, extraOptions);
    str = [codeBefore(UD), str ,codeAfter(UD)];
end
function [UD] = sanitizeUserData(UD);
    % make sure UD is a struct
    % it contains the right fields (or initialize them to empty defaults)
    % etc. etc.
end
```

That way the checks are concentrated and do not make the code more complex at the current generators. The only thing we have to account for, is that all `actualHandlers` only account for `extraAxisOptions` at the end.

Currently, I don't care anymore about the downside that this requires knowledge of our users of the HG object tree: with great power comes great responsibility.

**Comment by PeterPablo** [comment]
:+1: exactly the approach I would go for. We (only) need to make sure that the user supplied options clash with the ones that are generated by m2t. It probably makes sense to give the user's settings precedence.

**Comment by egeerardyn** [comment]
@PeterPablo : 
Indeed, I would also say that user-provided options should get precedence. Otherwise, the user shouldn't have provided those options in the first place, again "with great power comes great responsibility".

I think that if we make sure that `extraOptions` is handled at the very end of each `actualHandler` using `opts_append` (or was it `opts_add`, the one without any checking for duplicate values), that we should get the behavior we want.

**Comment by egeerardyn** [comment]
With #770, I've had a closer look at the code and the approach I put forth above seems very feasible for most of the objects. The only exceptional cases that I see are `axes` and `figure` objects. Since those are handled differently in `matlab2tikz`, handling those will require more work.

In any case: since we have some of the options available for those already, I don't think that should stand in the way of everything. In a first step, we can just implement the simple handlers and in a next step, we can take care of `axes` and `figure`.

For me, this is a feature that can go into `m2t 2.0`. The changes to the code are actually not that difficult, once we can provide nice functions to make the right data structures to store the extra user-defined code.

What needs to be stored with a handle's `UserData`, is:
 - code to be put before the element `<A>`
 - options to add to the element `<B>`
 - code to be put in the element, before the actual content `<C>`
 - code to be put in the element, after the actual content `<D>`
 - code to be put after the element `<E>`.

I wouldn't bother with adding "comments" to these in the beginning. A user can do so themself and adding it afterwards is just a few fields away.

So, in a template (when element = `axis` and indentation just for clarity) this would look like:
```latex
\begin{tikzpicture}
  <A>
  \begin{axis}[..., <B>]
  <C>
    \addplot{...};
  <D>
  \end{axis}
  <E>
\end{tikzpicture}
```

In principle, `<A>` and `<E>` can be added without modifying the actual handlers. `<B:D>`, however, require changing every handler a bit.

**Comment by andrew-d-king** [comment]
For what it's worth, I used a workaround that didn't involve editing the output of matlab2tikz.  My aim was to put extra code at the end of the {axis}, but instead I put it in the beginning of the {axis} by putting ']<text>%' at the end of my extraAxisOptions parameter.  This closes off the axis options "prematurely" and I was able to specify an annotation in the axis.  Unfortunately the annotation was under the data instead of over the data, but I got around this using pgf layers.

**Comment by egeerardyn** [comment]
@andrew-d-king That's a very sneaky work-around, I like it. Thanks for sharing!

For everybody else: I have a part of the implementation of this feature in a private branch of mine. It's not yet functional, but I'm halfway.
